### PR TITLE
add a timeout on the main loop, exit if the domain has crashed

### PR DESCRIPTION
**Pull request by Wenzel**
- add a timeout on queue.get() in the main loop
- add try catch to catch get_event exceptions if the ioctl failed

### CODE REVIEW

File: nitro/nitro.py [hunk 1]
```diff
@@ -90,28 +91,37 @@ def listen(self):
...
+                e = NitroEvent(nitro_raw_ev, vcpu_io.vcpu_nb)
+                # put the event in the queue
+                # and wait for the event to be processed, when the main thread will set the continue_event
+                item = (e, continue_event)
+                queue.put(item)
+                continue_event.wait()
+                # reset continue_event
+                continue_event.clear()
+                vcpu_io.continue_vm()
+            except ValueError as e:
```
Comment (noxdafox) [inline nitro/nitro.py]: Which line raises value error?

Comment (Wenzel) [inline nitro/nitro.py]: `vcpu_io.get_event()` and `vcpu_io.continue_vm()`

Comment (noxdafox) [inline nitro/nitro.py]: Then I'd wrap only those two methods.

From Robert C. Martin in 'Clean Code':

> if the keyword 'try' exists in a function, it should be the very first word in the function and that there should be nothing after the catch/finally blocks.

Another good reading in merit:
https://realpython.com/blog/python/the-most-diabolical-python-antipattern/

File: nitro/nitro.py [hunk 2]
```diff
@@ -90,28 +91,42 @@ def listen(self):
...
+                if not self.domain.isActive():
+                    self.stop_request.set()
+            else:
+                # remember last continue_event for stop_listen()
+                self.current_cont_event = continue_event
+                yield event
+                continue_event.set()
+
+        # raise listen_vcpu exceptions if any
+        [f.result() for f in self.futures]
```
Comment (noxdafox) [inline nitro/nitro.py]: This is slower in this case as you are first building a list and then GC it.

A normal for loop is faster and more readable in this case.

File: nitro/nitro.py [hunk 3]
```diff
@@ -90,28 +91,42 @@ def listen(self):
...
-            continue_event.wait()
-            # reset continue_event
-            continue_event.clear()
-
-            vcpu_io.continue_vm()
+            try:
+                nitro_raw_ev = vcpu_io.get_event()
+            except ValueError as e:
+                logging.debug(str(e))
+            else:
```
Comment (noxdafox) [inline nitro/nitro.py]: You can also omit the else.

File: nitro/nitro.py (hunk 2 above)
Comment (noxdafox) [inline nitro/nitro.py]: Pythonic code.

```
for future in self.futures:
    if future.exception() is not None:
        raise future.exception()
```

**Pull request merged**